File: indicators-back/app/engine/periodo.py
"""Period calculation helper — shared between routers.

Extracted from resultados.py so the SQL preview endpoint can reuse
the same period calculation logic without circular imports.
"""

from calendar import monthrange
from datetime import date, timedelta


# Mapping from periodo literals to (inicio, fin) relative to today.
def calcular_periodo(periodo: str) -> tuple[date, date]:
    """Translate a periodo literal into a concrete (inicio, fin) date pair.

    Uses Python's calendar.monthrange for accurate month boundaries and
    datetime.weekday() (Monday=0) for week boundaries.

    Raises ValueError for unknown periodo literals.
    """
    hoy = date.today()

    if periodo == "mes_actual":
        inicio = hoy.replace(day=1)
        return inicio, hoy

    if periodo == "trimestre_actual":
        trimestre = (hoy.month - 1) // 3
        inicio_mes = trimestre * 3 + 1
        inicio = hoy.replace(month=inicio_mes, day=1)
        return inicio, hoy

    if periodo == "semestre_actual":
        inicio_mes = 1 if hoy.month <= 6 else 7
        inicio = hoy.replace(month=inicio_mes, day=1)
        return inicio, hoy

    if periodo == "anual_actual":
        inicio = hoy.replace(month=1, day=1)
        return inicio, hoy

    raise ValueError(f"Periodo desconocido: {periodo}")

File: src/engine/periodo.ts
/**
 * Period calculation helper — shared between routers.
 *
 * Extracted so the SQL preview endpoint can reuse the same period
 * calculation logic without circular imports.
 */

/**
 * Calculate the current calendar month's boundaries in UTC.
 *
 * Returns [inicio, fin] where:
 * - inicio: first day of the current month at 00:00 UTC
 * - fin: today at 00:00 UTC (for real-time calculation) or last day of month
 *
 * Use `mes_referencia` (inicio) as the canonical month identifier
 * when persisting results.
 */
export function calcularMesActual(): { inicio: Date; fin: Date; mes_referencia: Date } {
  const hoy = todayUTC();
  const inicio = new Date(Date.UTC(hoy.getUTCFullYear(), hoy.getUTCMonth(), 1));
  return { inicio, fin: hoy, mes_referencia: inicio };
}

/**
 * Calculate the boundaries for a specific month given its first day.
 * Used when recalculating historical months.
 */
export function calcularMesEspecifico(
  anio: number,
  mes: number, // 1-indexed (January = 1)
): { inicio: Date; fin: Date; mes_referencia: Date } {
  const inicio = new Date(Date.UTC(anio, mes - 1, 1));
  const fin = new Date(Date.UTC(anio, mes, 0)); // last day of month
  return { inicio, fin, mes_referencia: inicio };
}

/**
 * Return today's date in UTC, with time set to 00:00:00.
 */
function todayUTC(): Date {
  const now = new Date();
  return new Date(Date.UTC(now.getUTCFullYear(), now.getUTCMonth(), now.getUTCDate()));
}
